File: python/rafter_cli/core/command_interceptor.py
"""Command interception and risk assessment."""
from __future__ import annotations

import re
from dataclasses import dataclass

from .audit_logger import AuditLogger
from .config_manager import ConfigManager
from .risk_rules import (
    is_chained_command,
    assess_command_risk,
    match_critical_pattern,
    sanitize_command_for_matching,
)


@dataclass
class CommandEvaluation:
    command: str
    risk_level: str  # low | medium | high | critical
    allowed: bool
    requires_approval: bool
    reason: str | None = None
    matched_pattern: str | None = None


class CommandInterceptor:
    def __init__(self) -> None:
        self._config = ConfigManager()
        self._audit = AuditLogger()

    def evaluate(self, command: str) -> CommandEvaluation:
        risk_level = self._assess_risk(command)

        # Unconditional hard-block: catastrophic destructive commands (rm -rf /,
        # fork bombs, disk wipes, mkfs, …) are NEVER allowed, regardless of the
        # configured policy — or its absence. Security must not depend on a
        # policy being present or on the chosen mode (even allow-all / a custom
        # deny-list cannot opt out of these).
        if risk_level == "critical":
            return CommandEvaluation(
                command=command,
                risk_level="critical",
                allowed=False,
                requires_approval=False,
                reason="Matches built-in blocked pattern (critical destructive command)",
                matched_pattern=match_critical_pattern(command) or "builtin:critical-destructive",
            )

        cfg = self._config.load_with_policy()
        policy = cfg.agent.command_policy

        # Check blocked patterns.
        #
        # A deny-list match denies — that is what a deny-list is for — but it
        # must NOT rewrite the command's risk. Reporting every deny-list hit as
        # "critical" made the hook tell users a `gh pr create` was an
        # irreversible system-damage command. The assessed risk is reported as
        # assessed; the genuinely unconditional hard-blocks are the
        # CRITICAL_PATTERNS handled above, and the default deny-list is exactly
        # that set.
        for pattern in policy.blocked_patterns:
            if self._matches(command, pattern):
                return CommandEvaluation(
                    command=command,
                    risk_level=risk_level,
                    allowed=False,
                    requires_approval=False,
                    reason=f"Matches blocked pattern: {pattern}",
                    matched_pattern=pattern,
                )

        # Check the positive allowlist. Deliberately AFTER blocked_patterns and
        # BEFORE require_approval: a deny rule always wins, and an allow rule's
        # whole job is to suppress the approval prompt for a known-safe command.
        #
        # Two guards keep an allowlist from becoming a hole in the guard rail:
        #
        #   - A ``critical`` command is never allowlistable. NOTE: this guard is
        #     currently UNREACHABLE — evaluate() hard-blocks critical above,
        #     before this loop — and a mutation sweep proved it: deleting the
        #     guard leaves every test green. It is kept as defence in depth,
        #     because it becomes the only protection the day that early block is
        #     narrowed. Do not write a test claiming to exercise it; such a test
        #     passes with the guard deleted.
        #   - A match does not apply when the command holds more than one
        #     statement. Patterns are unanchored by request, so without this
        #     "^git push" would wave through ``rm -rf / && git push`` — or, via
        #     the newline the original regex missed, anything on a second line.
        allowed = getattr(policy, "allowed_patterns", None) or []
        if not isinstance(allowed, list):
            # Defence in depth behind the validator: iterating a str yields
            # CHARACTERS, and a leading "^" then matches every command.
            allowed = []
        for pattern in allowed:
            if not self._matches(command, pattern):
                continue
            if is_chained_command(command):
                # Fall through to normal classification rather than allowing.
                break
            if risk_level == "critical":
                break
            return CommandEvaluation(
                command=command,
                risk_level="low",
                allowed=True,
                requires_approval=False,
                reason=f"Matches allowed pattern: {pattern}",
                matched_pattern=pattern,
            )

        # Check approval patterns
        for pattern in policy.require_approval:
            if self._matches(command, pattern):
                return CommandEvaluation(
                    command=command,
                    risk_level=risk_level,
                    allowed=False,
                    requires_approval=True,
                    reason=f"Matches approval pattern: {pattern}",
                    matched_pattern=pattern,
                )

        # Policy mode. `risk_level` is the assessment made above — critical
        # already returned, so it is high/medium/low here.
        if policy.mode == "approve-dangerous" and risk_level == "high":
            return CommandEvaluation(
                command=command,
                risk_level=risk_level,
                allowed=False,
                requires_approval=True,
                reason="High risk command requires approval",
            )

        return CommandEvaluation(
            command=command,
            risk_level=risk_level,
            allowed=True,
            requires_approval=False,
        )

    def log_evaluation(self, evaluation: CommandEvaluation, action_taken: str) -> None:
        self._audit.log_command_intercepted(
            evaluation.command,
            evaluation.allowed,
            action_taken,
            evaluation.reason,
        )

    # ------------------------------------------------------------------

    @staticmethod
    def _matches(command: str, pattern: str) -> bool:
        """Match a command against a policy pattern.

        Matching runs against the SANITIZED command line, not the raw string: the
        policy patterns describe commands, so quoted text a command merely
        consumes as data (a commit message, a PR body) must not match them, while
        text a shell or eval wrapper executes (`bash -c "…"`) must. See
        `sanitize_command_for_matching`.
        """
        target = sanitize_command_for_matching(command)
        try:
            return bool(re.search(pattern, target, re.IGNORECASE))
        except re.error:
            return pattern.lower() in target.lower()

    @staticmethod
    def _assess_risk(command: str) -> str:
        return assess_command_risk(command)

File: node/src/core/command-interceptor.ts
import { ConfigManager } from "./config-manager.js";
import { AuditLogger } from "./audit-logger.js";
import {
  assessCommandRisk,
  matchedCriticalPattern,
  sanitizeCommandForMatching,
  CommandRiskLevel,
  CHAIN_OPERATORS,
} from "./risk-rules.js";

export type { CommandRiskLevel } from "./risk-rules.js";

export interface CommandEvaluation {
  command: string;
  riskLevel: CommandRiskLevel;
  allowed: boolean;
  requiresApproval: boolean;
  reason?: string;
  matchedPattern?: string;
}

export class CommandInterceptor {
  private config: ConfigManager;
  private audit: AuditLogger;

  constructor() {
    this.config = new ConfigManager();
    this.audit = new AuditLogger();
  }

  /**
   * Evaluate if a command should be allowed
   */
  evaluate(command: string): CommandEvaluation {
    const riskLevel = this.assessRisk(command);

    // Unconditional hard-block: catastrophic destructive commands (rm -rf /,
    // fork bombs, disk wipes, mkfs, …) are NEVER allowed, regardless of the
    // configured policy — or its absence. Security must not depend on a policy
    // being present (the default config may be missing one) or on the chosen
    // mode (even allow-all / a custom deny-list cannot opt out of these).
    if (riskLevel === "critical") {
      return {
        command,
        riskLevel,
        allowed: false,
        requiresApproval: false,
        reason: "Matches built-in blocked pattern (critical destructive command)",
        matchedPattern: matchedCriticalPattern(command) ?? "builtin:critical-destructive"
      };
    }

    const cfg = this.config.loadWithPolicy();
    const policy = cfg.agent?.commandPolicy;

    if (!policy) {
      // No policy configured — fall back to safe built-in defaults rather than
      // allow-all: high-risk commands still require approval.
      if (riskLevel === "high") {
        return {
          command,
          riskLevel,
          allowed: false,
          requiresApproval: true,
          reason: "High risk command requires approval"
        };
      }
      return {
        command,
        riskLevel,
        allowed: true,
        requiresApproval: false
      };
    }

    // Check blocked patterns (always block).
    //
    // A deny-list match denies — that is what a deny-list is for — but it must
    // NOT rewrite the command's risk. Reporting every deny-list hit as
    // "critical" made the hook tell users a `gh pr create` was an irreversible
    // system-damage command. The assessed risk is reported as assessed; the
    // genuinely unconditional hard-blocks are the CRITICAL_PATTERNS handled
    // above, and the default deny-list is exactly that set.
    for (const pattern of policy.blockedPatterns) {
      if (this.matchesPattern(command, pattern)) {
        return {
          command,
          riskLevel,
          allowed: false,
          requiresApproval: false,
          reason: `Matches blocked pattern: ${pattern}`,
          matchedPattern: pattern
        };
      }
    }

    // Check the positive allowlist. Deliberately AFTER blockedPatterns and
    // BEFORE requireApproval: a deny rule always wins, and an allow rule's
    // whole job is to suppress the approval prompt for a known-safe command.
    //
    // Two guards keep an allowlist from becoming a hole in the guard rail:
    //
    //   - A `critical` command is never allowlistable. `rm -rf /`, a DB drop
    //     and wiping .git stay blocked whatever the config says.
    //   - A match does not apply when the command contains a chain operator.
    //     Patterns are unanchored by request, so without this "^git push"
    //     would wave through `rm -rf / && git push`. This mirrors the same
    //     disqualification SAFE_PREFIX already carries in risk-rules.ts.
    for (const pattern of policy.allowedPatterns ?? []) {
      if (!this.matchesPattern(command, pattern)) continue;

      if (CHAIN_OPERATORS.test(command)) {
        // Fall through to normal classification rather than allowing.
        break;
      }
      if (this.assessRisk(command) === "critical") {
        break;
      }
      return {
        command,
        riskLevel: "low",
        allowed: true,
        requiresApproval: false,
        reason: `Matches allowed pattern: ${pattern}`,
        matchedPattern: pattern
      };
    }

    // Check approval patterns
    for (const pattern of policy.requireApproval) {
      if (this.matchesPattern(command, pattern)) {
        return {
          command,
          riskLevel,
          allowed: false,
          requiresApproval: true,
          reason: `Matches approval pattern: ${pattern}`,
          matchedPattern: pattern
        };
      }
    }

    // Check policy mode. `riskLevel` is the assessment made at the top of
    // evaluate() — critical already returned, so it is high/medium/low here.
    if (policy.mode === "approve-dangerous" && riskLevel === "high") {
      return {
        command,
        riskLevel,
        allowed: false,
        requiresApproval: true,
        reason: `High risk command requires approval`
      };
    }

    // deny-list / allow-all / unknown mode: not blocked, not approval-gated.
    return {
      command,
      riskLevel,
      allowed: true,
      requiresApproval: false
    };
  }

  /**
   * Log command evaluation result
   */
  logEvaluation(evaluation: CommandEvaluation, actionTaken: "blocked" | "allowed" | "overridden"): void {
    this.audit.logCommandIntercepted(
      evaluation.command,
      evaluation.allowed,
      actionTaken,
      evaluation.reason
    );
  }

  /**
   * Match a command against a policy pattern.
   *
   * Matching runs against the SANITIZED command line, not the raw string: the
   * policy patterns describe commands, so quoted text a command merely consumes
   * as data (a commit message, a PR body) must not match them, while text a
   * shell or eval wrapper executes (`bash -c "…"`) must. See
   * `sanitizeCommandForMatching`.
   */
  private matchesPattern(command: string, pattern: string): boolean {
    const target = sanitizeCommandForMatching(command);
    try {
      const regex = new RegExp(pattern, "i");
      return regex.test(target);
    } catch {
      // If pattern is not valid regex, try case-insensitive substring match
      return target.toLowerCase().includes(pattern.toLowerCase());
    }
  }

  /**
   * Assess risk level of command
   */
  private assessRisk(command: string): CommandRiskLevel {
    return assessCommandRisk(command);
  }
}
